exonum/src/sandbox/consensus.rs

```diff
@@ -3156,6 +3254,164 @@ fn handle_block_response_tx_in_pool() {
+
+    let block = BlockBuilder::new(&sandbox)
+        .with_duration_since_sandbox_time(PROPOSE_TIMEOUT)
+        .with_txs_hashes(&[tx1.hash(), tx2.hash()])
+        .with_state_hash(&sandbox.compute_state_hash(&[tx1.raw().clone(), tx2.raw().clone()]))
+        .build();
+
+    let precommit_1 = Precommit::new(
+        VALIDATOR_1,
+        HEIGHT_ONE,
+        ROUND_ONE,
+        &propose.hash(),
+        &block.hash(),
+        sandbox.time().into(),
+        sandbox.s(VALIDATOR_1),
+    );
+    let precommit_2 = Precommit::new(
+        VALIDATOR_2,
+        HEIGHT_ONE,
+        ROUND_ONE,
+        &propose.hash(),
+        &block.hash(),
+        sandbox.time().into(),
+        sandbox.s(VALIDATOR_2),
+    );
+    let precommit_3 = Precommit::new(
+        VALIDATOR_3,
+        HEIGHT_ONE,
+        ROUND_ONE,
+        &propose.hash(),
+        &block.hash(),
+        sandbox.time().into(),
+        sandbox.s(VALIDATOR_3),
+    );
+
+    sandbox.recv(&Status::new(
+        &sandbox.p(VALIDATOR_3),
+        HEIGHT_TWO,
+        &block.hash(),
+        sandbox.s(VALIDATOR_3),
+    ));
+
+    sandbox.add_time(Duration::from_millis(BLOCK_REQUEST_TIMEOUT));
+    sandbox.send(
+        sandbox.a(VALIDATOR_3),
+        &BlockRequest::new(
+            &sandbox.p(VALIDATOR_0),
+            &sandbox.p(VALIDATOR_3),
+            HEIGHT_ONE,
+            sandbox.s(VALIDATOR_0),
+        ),
+    );
+    sandbox.recv(&tx1);
+    sandbox.recv(&tx2);
+
+    // Invalid transactions order.
+    sandbox.recv(&BlockResponse::new(
+        &sandbox.p(VALIDATOR_3),
+        &sandbox.p(VALIDATOR_0),
+        block.clone(),
+        vec![precommit_1, precommit_2, precommit_3],
+        &[tx2.hash(), tx1.hash()],
+        sandbox.s(VALIDATOR_3),
+    ));
+
+    sandbox.assert_state(HEIGHT_ONE, ROUND_ONE);
+}
+
+#[test]
+fn handle_block_response_with_invalid_precommits() {
+    let sandbox = timestamping_sandbox();
+
+    let tx = gen_timestamping_tx();
+
+    let propose = ProposeBuilder::new(&sandbox)
+        .with_duration_since_sandbox_time(PROPOSE_TIMEOUT)
+        .build();
+
+    let block1 = BlockBuilder::new(&sandbox)
+        .with_duration_since_sandbox_time(PROPOSE_TIMEOUT)
+        .with_tx_hash(&tx.hash())
+        .with_state_hash(&sandbox.compute_state_hash(&[tx.raw().clone()]))
+        .build();
+
+    let block2 = BlockBuilder::new(&sandbox)
+        .with_duration_since_sandbox_time(PROPOSE_TIMEOUT)
+        .build();
+
+    let precommit_1 = Precommit::new(
+        VALIDATOR_1,
+        HEIGHT_ONE,
+        ROUND_ONE,
+        &propose.hash(),
+        &block1.hash(),
+        sandbox.time().into(),
+        sandbox.s(VALIDATOR_1),
+    );
+    let precommit_2 = Precommit::new(
+        VALIDATOR_2,
+        HEIGHT_ONE,
+        ROUND_ONE,
+        &propose.hash(),
+        &block1.hash(),
+        sandbox.time().into(),
+        sandbox.s(VALIDATOR_2),
+    );
+    // Precommit with invalid block hash.
+    let precommit_3 = Precommit::new(
```

Comment:
Can you name this variable as `precommit_for_other_block` for example, or other more verbose name?